### PR TITLE
Refactor: Transform PipefyClient into Facade Pattern with Service Layer Architecture

### PR DESCRIPTION
**Pull request by adriannoes**
## Overview

This PR refactors the `PipefyClient` class from a monolithic ~488-line implementation into a modular architecture using **Facade Pattern + Service Layer Pattern**. The refactor maintains **100% backward compatibility** while significantly improving code organization, testability, and maintainability.

## Problem

The original `PipefyClient` class had grown too large and mixed multiple responsibilities:
- GraphQL client configuration
- Pipe and Card operations
- Embedded GraphQL queries
- Data transformation utilities

This monolithic structure made the codebase difficult to maintain, test, and extend.

## Solution

**Three-layer architecture:**

1. **Facade Layer** (`PipefyClient`): Maintains the exact same public API, delegates all operations to services
2. **Service Layer** (`PipeService`, `CardService`): Encapsulates domain-specific business logic
3. **Infrastructure Layer** (`BasePipefyClient`, `queries/`, `utils/`): Shared GraphQL client, queries, and utilities

## Key Benefits

- ✅ **Maintainability**: Code organized into focused, single-responsibility modules
- ✅ **Testability**: Services and utilities can be tested independently
- ✅ **Scalability**: Easy to add new operations by extending services
- ✅ **Backward Compatibility**: Existing code works without any changes

## Changes

**New Files:**
- `base_client.py` - Shared GraphQL client infrastructure
- `pipe_service.py` / `card_service.py` - Domain-specific services
- `queries/` - Extracted GraphQL operations
- `utils/` - Data transformation utilities
- Comprehensive test suite for new architecture

**Modified:**
- `client.py` - Transformed into pure facade (448 → 29 lines)
- `container.py` - Updated import path
- `test_pipefy_client.py` - Adapted for new architecture
- `README.md` - Added architecture overview

## Testing

- ✅ All existing tests pass
- ✅ New comprehensive test suite added (services, utilities, facade)
- ✅ Smoke test validates architecture integrity
- ✅ Backward compatibility verified

## Migration

**No migration needed** - This is a pure refactor with zero breaking changes. Existing code continues to work as-is:

```python
from pipefy_mcp.services.pipefy import PipefyClient

client = PipefyClient()
pipe = await client.get_pipe(pipe_id=123)
card = await client.create_card(pipe_id=123, fields={...})
```

### CODE REVIEW

**Comment by adriannoes** [comment]
## Code Review Follow-up

Added improvements based on code review:

- **Type consistency**: Updated to modern Python syntax (`list[str]`, `Client | None`)
- **Test coverage**: Added edge case test for [create_card](cci:1://file:///Users/adrianno.esnarriaga/Coding/pipefy-mcp-server/src/pipefy_mcp/services/pipefy/client.py:29:4-31:68) with empty dict
- **Documentation**: Added docstrings to all facade methods
- **Module exports**: Configured [__init__.py](cci:7://file:///Users/adrianno.esnarriaga/Coding/pipefy-mcp-server/src/pipefy_mcp/services/pipefy/__init__.py:0:0-0:0) files to export services, queries, and utilities
- **Code clarity**: Enhanced inline comments for legacy behavior

✅ All 54 tests passing

File: src/pipefy_mcp/services/pipefy/base_client.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from typing import Any
+
+from gql import Client
+from gql.transport.httpx import HTTPXAsyncTransport
+from httpx_auth import OAuth2ClientCredentials
+
+from pipefy_mcp.settings import settings
+
+
+class BasePipefyClient:
```
Comment (Danielmoraisg) [inline src/pipefy_mcp/services/pipefy/base_client.py]: **suggestion(non-blocking):** We could make this into an async context manager so we can batch many queries under the same connection like this:

```python
async with PipefyClient() as client:
    card = await client.get_card(123) 
    await client.update_card(123, title="X")
    await client.move_card_to_phase(123, 5) 
```

In order to do this we will need to define `__aenter__` and `__aexit__` methods that instantiates the session calling the client's `__aenter__` and `__aexit__` and sotring it on a `self._session`

For now we only execute one query, but this allows us to make more complex tools in the future

PS: if you want to also keep the current behavior to avoid having to use a context manager everytime, you can check on the `execute_query` if a session already exists and if not keep current implementation

File: src/pipefy_mcp/services/pipefy/base_client.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        self, schema: str | None = None, client: Client | None = None
+    ) -> None:
+        """Create a base client.
+
+        Args:
+            schema: Optional schema string to pass to `gql.Client`.
+            client: Optional pre-built `gql.Client` to reuse (preferred for shared wiring).
+        """
+
+        self.client: Client = client or self._create_client(schema)
```
Comment (Danielmoraisg) [inline src/pipefy_mcp/services/pipefy/base_client.py]: **suggestion:** I found it misleading to silently ignore the schema when both are passed, I'd suggest one of two ways to improve this:

1 - Create a separate gql factory component that handles the client creation and receives a `schema | None` and decides to fetch_schema_from_transport or not (my recomendation)

2 - Create multiple signatures with `@overload`  where one receives schema and the other the client. On the implementation itself check if bothe were passed

File: tests/services/test_pipefy_client.py [hunk 3]
```diff
@@ -696,19 +686,157 @@ async def test_update_card_incremental_mode_value_format_conversion():
...
 @pytest.mark.unit
 def test_convert_values_to_camel_case_missing_field_id():
-    """Test _convert_values_to_camel_case raises ValueError when field_id is missing."""
-    client = PipefyClient.__new__(PipefyClient)
-
+    """Test convert_values_to_camel_case raises ValueError when field_id is missing."""
     values = [{"value": "test"}]  # Missing field_id
     with pytest.raises(ValueError, match="missing required 'field_id' key"):
-        client._convert_values_to_camel_case(values)
+        convert_values_to_camel_case(values)
```
Comment (Danielmoraisg) [inline tests/services/test_pipefy_client.py]: **suggestion:** Since we now have a separate file/suite of tests for these utils, WDYT of moving all these tests that previously used the _convert_values_to_camel_case to the `test_formatters.py`

**Review by Danielmoraisg — APPROVED** [review]
Great work on this refactor!

File: README.md [hunk 4]
```diff
@@ -103,6 +96,19 @@ To use this with Cursor, you need to register it as an MCP server in your settin
...
+
+```bash
+# Run all tests
+uv run pytest
+
+# Run with coverage report
+uv run pytest --cov=src/pipefy_mcp/services/pipefy --cov-report=term-missing
+
+# Run smoke tests for refactor validation
+uv run python scripts/smoke_test_refactor.py
```
[the remark this answers is not in the document]
Comment (adriannoes) [reply README.md]: Yeah, the tests in [test_pipefy_facade.py](cci:7://file:///Users/adrianno.esnarriaga/Coding/pipefy-mcp-server/tests/services/test_pipefy_facade.py:0:0-0:0) are **architectural tests** rather than traditional unit tests. Here's the key difference:

**[test_pipefy_facade.py](cci:7://file:///Users/adrianno.esnarriaga/Coding/pipefy-mcp-server/tests/services/test_pipefy_facade.py:0:0-0:0) (Architecture Tests):**
- Verify that [PipefyClient](cci:2://file:///Users/adrianno.esnarriaga/Coding/pipefy-mcp-server/src/pipefy_mcp/services/pipefy/client.py:12:0-76:85) is a **pure facade** that only delegates calls without modifying args/returns
- Ensure both services share the **same injected `gql.Client` instance** (single source of truth)
- These tests validate the **structural contracts** of the refactor (Facade Pattern + dependency injection)

**`tests/services/*` (Unit Tests):**
- Test **business logic** (field conversion, camelCase formatting, GraphQL variable shapes)
- Validate **domain behavior** (filtering required fields, error handling, mutation selection)
- Focus on **what the code does**, not how it's wired together

**Goal:** Keep the facade tests as a regression safety net for future refactors. They ensure that changes to internal services don't accidentally break the delegation pattern or leak new dependencies.

Think of it as: facade tests → "is it wired correctly?" vs unit tests → "does it work correctly?"

File: src/pipefy_mcp/services/pipefy/base_client.py (hunk 1 above)
Comment (adriannoes) [reply src/pipefy_mcp/services/pipefy/base_client.py]: Excellent suggestion for the future, I'll create an issue to implement it in a separate PR.

Comment (adriannoes) [reply src/pipefy_mcp/services/pipefy/base_client.py]: Issue link: https://github.com/gbrlcustodio/pipefy-mcp-server/issues/11

File: scripts/smoke_test_refactor.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    print("✅ Testing facade structure...")
+    try:
+        # Use patch to avoid OAuth dependencies
+        from unittest.mock import MagicMock, patch
+        from gql import Client
+        
+        mock_client_instance = MagicMock(spec=Client)
+        
+        with patch(
+            "pipefy_mcp.services.pipefy.base_client.BasePipefyClient._create_client",
```
Comment (Danielmoraisg) [inline scripts/smoke_test_refactor.py]: **Suggestion:** We generally avoid messing with private methods as they should be considered internal implementation.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (adriannoes) [reply README.md]: The [smoke_test_refactor.py](cci:7://file:///Users/adrianno.esnarriaga/Coding/pipefy-mcp-server/scripts/smoke_test_refactor.py:0:0-0:0) was a temporary script used during the refactor to validate architectural contracts (imports, facade structure, shared client injection).

Since the unit tests in [tests/services/test_pipefy_facade.py](cci:7://file:///Users/adrianno.esnarriaga/Coding/pipefy-mcp-server/tests/services/test_pipefy_facade.py:0:0-0:0) already cover these contracts with proper mocking, I've removed the smoke test script. 

It's no longer needed now that the refactor is complete.
